### PR TITLE
add soft fp8 feat(matmul and grouped matmul ops)

### PR DESCRIPTION
**Pull request by LinyuanLi0046**
This PR introduces Soft FP8 support by adding software-based FP8 computation paths with weight dequantization (FP8 → BF16).
- Implemented Soft FP8 MatMul operation with integrated weight dequantization.
- Implemented Soft FP8 Grouped MatMul operation with integrated weight dequantization.
- The implementation has been tested with the Qwen3-MoE-30B-A3B-FP8 and DeepSeekR1-0528 model.

### CODE REVIEW

File: tests/python/sgl_kernel_npu/test_fp8_w8a16_gmm.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import torch
```
Comment (randgun) [inline tests/python/sgl_kernel_npu/test_fp8_w8a16_gmm.py]: Need to add readme file.

Comment (LinyuanLi0046) [reply tests/python/sgl_kernel_npu/test_fp8_w8a16_gmm.py]: Added readme of matmul, gmm, bmm in /csrc/catlass/README.md.

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "tiling/platform/platform_ascendc.h"
+#include "defines.h"
+#include "torch_helper.h"
+#include "catlass_op_utils.h"
+#include "aclrtlaunch_catlass_fp8w8a16_gmm_bfloat16_t.h"
+
+namespace sglang {
+namespace npu_kernel {
+HOST_API at::Tensor softfp8_w8a16_grouped_matmul(const at::Tensor &mat1, const at::Tensor &mat2,
```
Comment (xdtbynd) [inline csrc/catlass/op_host/catlass_gmm_fp8.cpp:26]: Function name `softfp8_w8a16_grouped_matmul` does not match the invoked kernel name `catlass_fp8w8a16_gmm_bfloat16_t`, which may cause confusion, unify naming style or add a comment clarifying the mapping.

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                                                 const at::Tensor &scale, const at::Tensor &groupList,
+                                                 const std::string &outDType)
+{
+    at::ScalarType scalar_type = mat1.scalar_type();
+
+    TORCH_CHECK(scalar_type == at::kBFloat16, "only support bf16");
```
Comment (xdtbynd) [inline csrc/catlass/op_host/catlass_gmm_fp8.cpp:33]: Only `mat1` dtype is checked (kBFloat16). The function name suggests activation is BF16 and weights are FP8, but no validation is performed on `mat2` dtype. Suggest to add TORCH_CHECK to verify the dtype of `mat2`

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    TORCH_CHECK(mat1.dim() == 2, "x should be [M, K]");
+    TORCH_CHECK(mat2.dim() == 3, "weight should be [g, k, n]");
```
Comment (xdtbynd) [inline csrc/catlass/op_host/catlass_gmm_fp8.cpp:36]: Missing shape and type validation for `scale` and `groupList`. Invalid inputs could cause out-of-bounds access or incorrect computation in the kernel. Suggest to add input validation based on kernel expectations: e.g., dimensions of `scale`, validity of `groupList`.

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    uint32_t m = mat1.size(0);
+    uint32_t k = mat1.size(1);
+    uint32_t n = mat2.size(2);
+    uint32_t g = mat2.size(0);
+
+    void *x_ptr = mat1.data_ptr();
```
Comment (xdtbynd) [inline csrc/catlass/op_host/catlass_gmm_fp8.cpp:59]: Calls `data_ptr()` directly without checking tensor contiguity. Suggest to all `contiguous()`

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    void *group_list_ptr = groupList.data_ptr();
+
+    auto outputDataType = TypeStrToAclDtype(outDType);
+    at::Tensor output = GetOutputTensor({m, n}, AclDtypeToTorchDtype(outputDataType));
+    void *y_ptr = output.data_ptr();
+
+    aclrtStream stream = c10_npu::getCurrentNPUStream().stream(false);
+    uint32_t aicCoreNum = platform_ascendc::PlatformAscendCManager::GetInstance()->GetCoreNumAic();
+    uint32_t workspace_size = 4 * 256 * 256 * sizeof(outputDataType) * aicCoreNum;
+    auto workspace_tensor =
```
Comment (xdtbynd) [inline csrc/catlass/op_host/catlass_gmm_fp8.cpp:71]: `workspace_tensor allocation` may fail. Check that `workspace_ptr` is not null.

File: csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+#include "catlass/gemm/block/block_swizzle.hpp"
+#include "catlass/gemm/dispatch_policy.hpp"
+#include "catlass/gemm/gemm_type.hpp"
+#include "catlass/layout/layout.hpp"
+#include "catlass/status.hpp"
+
+#include "../utils/gemm/kernel/gmm_wfp8a16.hpp"
+#include "../utils/gemm/tile/cast_fp8_to_bf16_with_scale.hpp"
+#include "../utils/gemm/block/block_mmad_pingpong_with_prologue_fp8_w8a16.hpp"
+#include "../utils/gemm/dispatch_policy_ext.hpp"
```
Comment (xdtbynd) [inline csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp:29]: Header includes are mixed (AscendC, Catlass, and local paths). Some paths like `"../utils/gemm/kernel/gmm_wfp8a16.hpp"` are relative and may break if the build directory structure changes.

File: csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+using namespace Catlass;
+using namespace matmul;
+
+extern "C" __global__ __aicore__ void catlass_fp8w8a16_gmm_bfloat16_t(GM_ADDR deviceA, GM_ADDR deviceB,
+                                                                      GM_ADDR deviceScale, GM_ADDR deviceGroupList,
+                                                                      GM_ADDR deviceC, GM_ADDR deviceWorkspace,
+                                                                      uint32_t g, uint32_t m, uint32_t n, uint32_t k)
+{
+    KERNEL_TASK_TYPE_DEFAULT(KERNEL_TYPE_MIX_AIC_1_2);
```
Comment (xdtbynd) [inline csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp:39]: Suggest to add a comment explaining why this specific task type is chosen

File: csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    using DispatchPolicy = Gemm::MmadAtlasA2PingPongWithPrologueBFP8<ENABLE_UNIT_FLAG>;
+    using PrologueA = void;
+    constexpr uint32_t computeLen = 8 * 1024;
+    using PrologueB = Gemm::Tile::TileCastFp8ToBf16WithScaleDequant<ArchTag, PrologueSrcType, PrologueDstType,
+                                                                    PrologueScaleType, computeLen>;
+    using TileCopy = Gemm::Tile::TileCopyWithPrologue<ArchTag, AType, BType, CType, PrologueA, PrologueB>;
+    using BlockMmadOpt =
+        Gemm::Block::BlockMmad<DispatchPolicy, L1TileShape, L0TileShape, AType, BType, CType, void, TileCopy>;
+
+    if (k > n) {
```
Comment (xdtbynd) [inline csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp:91]: `if (k > n) { ... } else { ... }` – Both branches are nearly identical, only differing in the last template argument of `BlockScheduler`. Suggest to factor out the common code into a macro or a helper function template with the scheduler as a template parameter.

File: tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+import sgl_kernel_npu
+import torch
+import torch_npu
+
+
+def anti_quant_fp8_to_bf16(weight, scale, G, K, N):
+    device = weight.device
+
+    scale_fp8_to_32 = torch.tensor(0x7B80, dtype=torch.uint16, device=device)
```
Comment (xdtbynd) [inline tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py]: Magic number, it needs a comment.

File: tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            i, :, :
+        ].to(torch.float32)
+        ret.append(c)
+    ret = torch.cat(ret, dim=0)
+    print(f"ret shape: {ret.shape}, {ret.dtype}")
+    return ret
+
+
+def compare_data_Wf8Abf16(g, m, n, k):
+    single_m = int(m / g)
```
Comment (xdtbynd) [inline tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py]: This assumes `m` is divisible by `g`. If not, the grouping will be wrong (the last group would have different size). The script does not check this.

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp (hunk 3 above)
Comment (LinyuanLi0046) [reply csrc/catlass/op_host/catlass_gmm_fp8.cpp:33]: accepted

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp (hunk 4 above)
Comment (LinyuanLi0046) [reply csrc/catlass/op_host/catlass_gmm_fp8.cpp:36]: accept

File: csrc/catlass/op_host/catlass_gmm_fp8.cpp (hunk 5 above)
Comment (LinyuanLi0046) [reply csrc/catlass/op_host/catlass_gmm_fp8.cpp:59]: Contiguous should be guaranteed at the time of invocation.

File: csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp (hunk 7 above)
Comment (LinyuanLi0046) [reply csrc/catlass/op_kernel/catlass_gmm_fp8_kernel.cpp:29]: Thanks for pointing this out; however, since the current structure is stable and thoroughly tested, I'd like to postpone this path optimization to our next routine maintenance to minimize regression risks.

File: tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py (hunk 11 above)
Comment (LinyuanLi0046) [reply tests/python/sgl_kernel_npu/test_gmm_wfp8a16.py]: accept

File: csrc/catlass/README.md [hunk 12]
```diff
@@ -76,3 +76,345 @@ for dtype in dtypes:
...
+scale_cols = (N + 127) // 128
+scale = torch.ones((g, scale_rows, scale_cols), dtype=torch.float32, device=device)
+
+c = torch.ops.npu.softfp8_w8a16_grouped_matmul(a, w, scale, groupList_t, "bf16")
+print(c.shape, c.dtype)  # (M, N), bf16
+```
+
+---
+
+# torch.ops.npu.softfp8_w8a16_batch_matmul
```
Comment (LinyuanLi0046) [inline csrc/catlass/README.md:317]: missed kernel softfp8_batchmatmul mentioned in doc

File: .gitmodules [hunk 13]
```diff
@@ -1,3 +1,7 @@
-[submodule "third-party/pto-isa"]
-	path = third-party/pto-isa
+[submodule "third_party/pto-isa"]
```
Comment (LinyuanLi0046) [inline .gitmodules:1]: add catlass to third_party dir, however, in main branch, pto-isa add submodule in dir third-party, shoud keep consistency with main branch.

File: csrc/catlass/README.md [hunk 14]
```diff
@@ -76,3 +76,345 @@ for dtype in dtypes:
...
+
+## Function Description | 功能描述
+
+### English:
+Catlass-based soft-FP8 (W8A16) matmul. Activations are BF16; weights are FP8 bits stored as int8 and are dequantized to BF16 with per block scales(block size is (128, 128)), then GEMM is executed and BF16 output is produced.
+
+This operator is applicable to Atlas A2/A3 and enables FP8 computation on Ascend 910B/C hardware that does not support native FP8 computation.
+
+### 中文:
+基于 Catlass 的 soft-FP8（W8A16）矩阵乘：输入激活为 BF16；权重以 int8 形式存储 FP8 bit，按照块用 scale 做反量化（块大小是(128, 128)），转换为 BF16，再进行 GEMM，输出为 BF16。
```
Comment (LinyuanLi0046) [inline csrc/catlass/README.md:90]: Some descriptions and examples in the documentation are not completely consistent with the parameter validation in the code.

Comment (LinyuanLi0046) [reply csrc/catlass/README.md:90]: fixed

File: csrc/catlass/README.md (hunk 12 above)
Comment (LinyuanLi0046) [reply csrc/catlass/README.md:317]: in another PR https://github.com/sgl-project/sgl-kernel-npu/pull/278

File: .gitmodules (hunk 13 above)
Comment (LinyuanLi0046) [reply .gitmodules:1]: fixed

**Review by iforgetmyname — APPROVED** [review]
